### PR TITLE
Using the rollout coordinates and stop gradient on representations for confidence head 

### PR DESCRIPTION
**Pull request by gitabtion**
This PR has 2 contributions:
**1. Using the rollout coordinates for confidence head**
As supplementary information section 4.1, 

> Several of the heads require predicted coordinates, therefore at training time we do a short rollout of the Diffusion Module from pure noise with 20 steps (Main Article Fig. 2c). No gradients are applied to this mini-rollout.

the predicted coordinates was generate by mini-rollout. On other hand, the denoised coordinates generate by diffusion training step is only one of the denoise step of the predict coordinates.

**2. Apply stop gradient on representations and predict structure.** 
As supplementary information section 4.3.5

> During training, a stop gradient is applied to the single and pair embeddings and predicted structure.

### CODE REVIEW

**Comment by lucidrains** [comment]
@gitabtion thank you Heng!

File: alphafold3_pytorch/alphafold3.py [hunk 1]
```diff
@@ -3361,17 +3362,31 @@ def forward(
...
+                mask = mask,
+                single_trunk_repr = single,
+                single_inputs_repr = single_inputs,
+                pairwise_trunk = pairwise,
+                pairwise_rel_pos_feats = relative_position_encoding,
+                residue_atom_lens = residue_atom_lens
+            )
 
             if self.packed_atom_repr:
                 pred_atom_pos = einx.get_at('b [m] c, b n -> b n c', denoised_atom_pos, residue_atom_indices)
```
Comment (lucidrains) [inline alphafold3_pytorch/alphafold3.py:3381]: @gitabtion hey Heng, i don't think you accounted for `pred_atom_pos` being overwritten here

Comment (gitabtion) [reply alphafold3_pytorch/alphafold3.py:3381]: As mentioned in section 4.1, the `x_pred` used by the Auxiliary heads consists of coordinates generated through mini-rollouts, which aligns with our understanding. The denoised coordinates produced during the diffusion training steps are intermediate states in the generation process and differ significantly from the final denoised coordinates generated during inference. On the other hand, the Confidence heads are designed to evaluate the confidence of the generated structures. Therefore, during the training of the Confidence heads, it is appropriate to use coordinates generated from complete sampling (rollout) as input features and compute labels such as pLDDT, PAE, PDE, etc., using these coordinates and the ground truth (which is yet to be implemented).

The utilization of rollout-generated coordinates to replace `pred_atom_pos` serves as input features for the Confidence heads. Additionally, due to the presence of symmetry in certain atoms or chains, we also need to implement a procedure to align and adjust the order of the ground truth's atoms or chains using coordinates generated through rollouts.

Comment (gitabtion) [reply alphafold3_pytorch/alphafold3.py:3381]: @lucidrains ah, I made a mistake! It should be named `denoised_atom_pos` instead of `pred_atom_pos`.

Comment (lucidrains) [reply alphafold3_pytorch/alphafold3.py:3381]: indeed, thanks!
